Initial Rebase implementation

jamill opened the pull request:
Getting closer to completion

**Includes:**
- Initiate a Rebase (merge strategy)
- Abort a Rebase (merge strategy)
- Continue a Rebase (merge strategy)
- Progress reporting
  - Before / After each rebase step
  - Checkout progress during each rebase step

**Does not include:**
- Interactive Rebase support, of any kind
- Support for rebases not using the merge strategy
- Ability to execute rebase "commands"

**TODO:**
- [x] Expand test coverage
  - [x] Tests for checkout progress
  - [x] Tests for FileConflictStrategy

**To Think About:**
- Ref inputs: I think it would be nice if methods such as merge, rebase could take in a "generic" object or interface that can be peeled to a commit, but still have information about where it came from. For instance, rebase takes 3 "ref" inputs. Instead of adding overloads for each permutation of ref input (commit, string, branch, tag...). I suspect this could be similar to the `git_annotated_commit` concept in libgit2.

Related issue: #714

nulltoken commented on LibGit2Sharp/IRepository.cs:
```diff
@@ -223,6 +223,22 @@ public interface IRepository : IDisposable
+        /// <summary>
+        /// Access to Rebase functionality.
+        /// </summary>
+        Rebase RebaseOperation { get; }
```
How do you foresee to use this?

nulltoken replied:
Found it. Not sure I'm a fan of having the `Rebase()` method in the main namespace and the other methods under `RebaseOperation`. How about exposing a separate `repo.Rebase` namespace (à la `repo.Diff`)?

We could replace `repo.Rebase()` with something like `repo.Rebase.Start()`, `Initiate()`, ...

jamill replied:
yes - that seems reasonable. Thinking about the larger picture, cherry-pick and revert both have a similar "sequenced" capability. If we put rebase in its own namespace, then we would we have `repo.CherryPick.Start(...)`, and `repo.Revert.Start(...)`? These (and Merge) all seem to be similar types of operations, would Merge also move into a `repo.Merge.Start(...)` type of structure? One option is that we could do this, and then have extension methods to expose these on the repository object directly?

nulltoken commented on LibGit2Sharp/RebaseOperation.cs:
```diff
@@ -0,0 +1,172 @@
+        /// No commit to cherry-pick. Run the given command and continue
+        /// if successful.
+        /// </summary>
+        Exec
```
Hmmm... What's the plan for this? A Callback?

nulltoken commented:
How does the code behave when

- One initiates a new Rebase() whereas one is already started?
- One tries to continue (or abort) when no rebase is in progress?

jamill commented:
> How does the code behave when...

The code should throw an exception in these cases. There is already a test case that covers initiating a rebase while one is already started [RebaseWhileAlreadyRebasingThrows()](https://github.com/libgit2/libgit2sharp/pull/964/files#diff-5a992064e68cb142d2863abc17cf25acR198). I can add test coverage to assert we throw in the other case as well.

jamill commented:
1. I moved the entry point for rebase into the `Rebase` namespace.

2. I added tests to assert the behavior when attempting to abort, continue a rebase when one is not in progress. This is currently returning a `NotFoundException` (as it can't find the rebase operation). This is a bit weird, we could return a specificly typed exception here.

rcorre commented on LibGit2Sharp/Rebase.cs:
```diff
@@ -0,0 +1,266 @@
+        /// <param name="committer"></param>
+        /// <param name="options"></param>
+        /// <returns>true if completed successfully, false if conflicts encountered.</returns>
+        public virtual RebaseResult Start(Branch branch, Branch upstream, Branch onto, Signature committer, RebaseOptions options)
```
In the long run, I think it would be more flexible to take a commit rather than a branch, at least for the `onto` argument. I often use `rebase -i some_commit_id` to clean up before a push. This could be added as an overload later though.

ethomson replied:
We probably don't want to take a bare commit here - these things end up influencing the messages that we produce in the reflog and the markers in the conflict files.  Do we have the equivalent of an `AnnotatedCommit` in libgit2sharp?

ethomson replied:
Sorry, to expand on that a little bit more:  I mean we probably shouldn't *only* take a commit id (but we should perhaps allow it).  Although a `Branch` is going to be convertible to a commit ID at some point, if you let libgit2 handle that conversion it will keep around the information of where that commit ID came from (as an "annotated commit").  Without that your conflict markers will look like:

    <<<< ab1dc99f...
    change on one side
    ===== ab1cdde
    common ancestor
    =====
    change on the other side
    >>>>> 9acca3a

Which is quite hard to read (I think) compared to having the branch names / etc.

rcorre replied:
Could it just take `GitAnnotatedCommitHandle`s as arguments? The branch args are getting converted into these anyways.

jamill replied:
Thanks @rcorre - GitAnnotatedCommitHandle is a lower level concept between lg2# and lg2 - it does not get exposed on the lg2# API. I agree with the idea that the API should be more flexible in the inputs it accepts (this is one of the items I had called out in my **To Think About** section in the PR description).

nulltoken commented:
Rebased this PR. Ah, the irony...

nulltoken commented on LibGit2Sharp/Core/Proxy.cs:
```diff
@@ -1601,6 +1602,189 @@ public static void git_patch_print(PatchSafeHandle patch, NativeMethods.git_diff
+
+            using (ThreadAffinity())
+            {
+                int result = NativeMethods.git_rebase_abort(rebase);
```
Shouldn't we `Ensure` here?

jamill replied:
We now `Ensure` the result - thanks for pointing this out!

Therzok commented on LibGit2Sharp.Tests/TestHelpers/Constants.cs:
```diff
@@ -8,6 +8,7 @@ public static class Constants
         public const string UnknownSha = "deadbeefdeadbeefdeadbeefdeadbeefdeadbeef";
         public static readonly Identity Identity = new Identity("A. U. Thor", "thor@valhalla.asgard.com");
         public static readonly Signature Signature = new Signature(Identity, new DateTimeOffset(2011, 06, 16, 10, 58, 27, TimeSpan.FromHours(2)));
+        public static readonly Signature Signature2 = new Signature("nulltoken", "emeric.fermas@gmail.com", DateTimeOffset.Parse("Wed, Dec 14 2011 08:29:03 +0100"));
```
Should we have proper emails in plain text here? What if @nulltoken gets spammed? :D

jamill replied:
yeah, we have these emails in several places. If we want to clean them, we should probably scrub these everywhere.

dahlbyk replied:
Incoming `filter-branch` and re-SHAing of the test repos. :wink:

Therzok commented on LibGit2Sharp/Core/Proxy.cs:
```diff
@@ -1601,6 +1602,190 @@ public static void git_patch_print(PatchSafeHandle patch, NativeMethods.git_diff
+            RebaseSafeHandle rebase,
+            long index)
+        {
+            Debug.Assert(index >= 0);
```
Should this line still be here? Just curious.

jamill replied:
I think the Debug Assert is OK to leave in.

whoisj replied:
I'd vote for more Debug Asserts in the code. There's a lot of places we make assumptions and never assert then.

Therzok replied:
My only reasoning is: Why is it a Debug only assert? Release builds will have it stripped off.

whoisj replied:
Yes, given that it is public, it likely should ensure and no assert.

Therzok replied:
I kept forgetting that Proxy is internal.

Therzok commented on LibGit2Sharp/Core/Proxy.cs:
```diff
@@ -1601,6 +1602,190 @@ public static void git_patch_print(PatchSafeHandle patch, NativeMethods.git_diff
+                finally
+                {
+                    committerHandle.Dispose();
+                    committerHandle = null;
```
Question/nit: Why these try{}finally{} instead of using? Also, no need for explicit null setting. What if creating an author handle failed and authorHandle is null?

This piece needs cleaning.

jamill replied:
We are creating several safe handles here (and other places with this PR) - I didn't want to nest usings deeply in this case (especially when there is logic in determing how the safe handles are created).

`authorHandle` is SafeDisposed, which handles the null case.

Therzok commented on LibGit2Sharp/Core/Proxy.cs:
```diff
@@ -3461,5 +3646,28 @@ internal static int ConvertResultToCancelFlag(bool result)
+        /// <returns></returns>
+        public static long ConvertToLong(this UIntPtr input)
+        {
+            ulong ulongValue = (ulong)input;
```
I think this got added by @nulltoken in vNext.

Therzok commented on LibGit2Sharp/Rebase.cs:
```diff
@@ -0,0 +1,271 @@
+            }
+            finally
+            {
+                branchRefPtr.SafeDispose();
```
Same comment as above. We're explicitly nulling out local variables.

Therzok commented on LibGit2Sharp/Rebase.cs:
```diff
@@ -0,0 +1,271 @@
+            }
+            finally
+            {
+                rebase.SafeDispose();
```
Another nit as above.

jamill commented:
@Therzok I have some changes for moving more of the code into using `using`s - will push it up shortly

Therzok commented:
:heart:

Therzok commented on LibGit2Sharp.Tests/RebaseFixture.cs:
```diff
@@ -0,0 +1,574 @@
+
+                // Verify the chain of commits that resulted from the rebase.
+                Commit expectedParent = expectedOntoCommit;
+                foreach(CompletedRebaseStepInfo stepInfo in PostRebaseResults)
```
Nit: missing spaaaaace.

Therzok commented on LibGit2Sharp.Tests/RebaseFixture.cs:
```diff
@@ -0,0 +1,574 @@
+
+            public bool WasPatchAlreadyApplied { get; set; }
+
+                        public override string ToString()
```
Unindent by 8 spaces.

Therzok commented on LibGit2Sharp.Tests/RebaseFixture.cs:
```diff
@@ -0,0 +1,574 @@
+                    return true;
+                }
+
+                if ((x == null && y != null ) ||
```
I might be a devil, but there's an extra space before each parenthesis.

ThomasBarnekow commented on LibGit2Sharp/Core/Proxy.cs:
```diff
@@ -1643,6 +1645,174 @@ public static void git_patch_print(PatchSafeHandle patch, NativeMethods.git_diff
+            Ensure.ArgumentNotNull(signature, "signature");
+
+            using (ThreadAffinity())
+            using (var signatureHandle = signature.BuildHandle())
```
Having had a lengthy discussion with @Therzok about explicitly declaring variables if the type can not be inferred from the righthand side, should we use ```var``` here? 

At least I would not know which type ```signature.BuildHandle()``` returns from looking at just this line. While I understand that it returns a signature handle and that the ```using``` is only there to properly dispose of that handle, I wanted to point that out to make sure we are consistent. 

I'd personally be fine with the ```var``` in this construct, though.

nulltoken commented on LibGit2Sharp/Rebase.cs:
```diff
@@ -0,0 +1,219 @@
+        /// No commit to cherry-pick. Run the given command and continue
+        /// if successful.
+        /// </summary>
+        Exec
```
How do we intend to use this?

jamill replied:
We do not currently support this - libgit2 does not support interactive rebase yet. We will have to think about how we want to handle this type of operation moving forward.

nulltoken replied:
So do we need to expose it now? Meaning, couldn't we just expose it when we actually know how to handle it?

jamill replied:
we could - but these are just reflecting the values exposed by lg2 at the moment. I don't feel too strongly about it, either way.

nulltoken replied:
Let's hide those for now, please.

nulltoken commented on LibGit2Sharp/Rebase.cs:
```diff
@@ -0,0 +1,219 @@
+
+            options = options ?? new RebaseOptions();
+
+            if (this.repository.Info.CurrentOperation != CurrentOperation.None)
```
Should we guard against BareRepositories as well?

jamill replied:
We could guard against it here, or let the native layer do these validations... I will add a check here for now.

nulltoken commented on LibGit2Sharp/Rebase.cs:
```diff
@@ -0,0 +1,219 @@
+        /// </summary>
+        /// <param name="committer">The <see cref="Signature"/> of who added the change to the repository.</param>
+        /// <param name="options">The <see cref="RebaseOptions"/> that specify the commit behavior.</param>
+        public virtual RebaseResult Continue(Signature committer, RebaseOptions options)
```
Do we still need a `Signature`? Or can we switch to `Identity`?

nulltoken replied:
What happens if we `Continue` a non already started rebase?

jamill replied:
> Do we still need a  Signature ? Or can we switch to  Identity ?

I updated to use `Identity`.

> What happens if we  Continue  a non already started rebase?

This will throw (behavior is coverend in the  RebaseOperationsWithoutRebasingThrow).

nulltoken commented on LibGit2Sharp/Rebase.cs:
```diff
@@ -0,0 +1,219 @@
+        /// <summary>
+        /// Abort the rebase operation.
+        /// </summary>
+        public virtual void Abort()
```
How does this behave if no rebase is in progress? NoOp?

jamill replied:
This will throw. I covered this behavior in the `RebaseOperationsWithoutRebasingThrow` test

nulltoken commented on LibGit2Sharp/Rebase.cs:
```diff
@@ -0,0 +1,219 @@
+        /// </summary>
+        public virtual void Skip()
+        {
+            throw new NotImplementedException();
```
Are we willing to ship as is? Couldn't we not expose this method until it's ready?

nulltoken commented on LibGit2Sharp/Rebase.cs:
```diff
@@ -0,0 +1,219 @@
+        /// <summary>
+        /// The info on the current step.
+        /// </summary>
+        public virtual RebaseStepInfo GetCurrentStepInfo()
```
I'm not sure about `Step`. The returned data deals more with the state of the `Process` rather than a specific step.

jamill replied:
This returns a `RebaseStepInfo` object, which also contains the total number of steps. I had thought about separating out the total step count, but hadn't done so. I am open other suggestions for the name.

carlosmn replied:
Looking into it a bit more, I share @nulltoken 's concern about it being more than the step, it's more like "the current step, and by the way, here's a bit of information about the overall rebase operation, 'cos that way we have it".

I would expect each step to describe itself, and it the caller wants to figure out how far along they are in the rebase, they can ask the operation itself. Unless I'm missing some use-case where this is not feasible.

dahlbyk replied:
It's worth noting that a paused rebase may resume with more or fewer steps remaining if its todo list is edited.

nulltoken replied:
Interesting. @jamill Do we support that yet?

ethomson replied:
I suppose that should be supported  If we're deficient here, it's libgit2 that's deficient.  I would expect this to work, libgit2 doesn't keep state between steps (because that's impractical for interop with git core, not for this reason).

jamill replied:
I haven't tried this explicitly - but I agree with @ethomson's assessment: I would expect this to work. LibGit2Sharp does not cache anything - it queries into libgit2 each time.

dahlbyk replied:
Are these scenarios (add/remove step mid-rebase) we could/should test?

jamill replied:
I think these are scenarios we could include later on to flesh out the overall feature. At this point, the APIs do not support interactive rebase, and so there is not really a way to "pause" the rebase (outside of conflicts). I can create an issue to track the testing of this scenario.

dahlbyk replied:
:+1:

ethomson replied:
Right now we don't support interactive rebase, we only support `git rebase --merge`.  It's very challenging (but not impossible) for a user to change the "script" that rebase is performing.  So it would be hard to test at the moment.  We should have a test here when we implement interactive rebase.

nulltoken commented on LibGit2Sharp/RebaseResult.cs:
```diff
@@ -0,0 +1,76 @@
+    /// <summary>
+    /// The status of the rebase.
+    /// </summary>
+    public enum RebaseStatus
```
Couldn't we use past tense for every term (`Completed`, `Conflicted` (not sure about this), `StopRequested`)?

nulltoken replied:
Or... keep the two first ones and only rename `RequestedStop` into something else ?

jamill replied:
We are a bit inconsistent here in general - `RevertStatus` has `Reverted` and `Conflicts`. `MergeStatus` seems to be mostly present tense. I am open to suggestions... For now, I will go with `Stop`, but that reads funny to me.

nulltoken replied:
Let's :ship: it, then.

nulltoken commented on LibGit2Sharp/RebaseStepInfo.cs:
```diff
@@ -0,0 +1,54 @@
+        /// <summary>
+        /// Command to execute, if any.
+        /// </summary>
+        public virtual string Exec { get; private set; }
```
Similarly do we need to expose this yet?

nulltoken commented on LibGit2Sharp/Core/Proxy.cs:
```diff
@@ -1643,6 +1645,192 @@ public static void git_patch_print(PatchSafeHandle patch, NativeMethods.git_diff
+            IntPtr ptr = NativeMethods.git_rebase_operation_byindex(rebase, ((UIntPtr)index));
+            GitRebaseOperation operation = ptr.MarshalAs<GitRebaseOperation>();
+
+            // Workaround until 92e87dd74 from libgit2 is consumed by LibGit2#
```
Could you please update to the latest tip (or a commit without too many breaking changes) and remove this workaround?

nulltoken commented on LibGit2Sharp/Core/Proxy.cs:
```diff
@@ -1643,6 +1645,192 @@ public static void git_patch_print(PatchSafeHandle patch, NativeMethods.git_diff
+                // If successsful, then marshal native struct to managed struct.
+                operation = ptr.MarshalAs<GitRebaseOperation>();
+
+                // Workaround until 92e87dd74 from libgit2 is consumed by LibGit2#
```
Same here?

jamill commented:
@nulltoken tests are green, rebased on top of vNext.

jamill commented:
@nulltoken - anything else here? :smile:

nulltoken commented:
:shipit: :question: 

/cc @Therzok @dahlbyk

Therzok commented:
LGTM! :rocket: :ship: :it:

nulltoken commented:
@jamill :100: :bangbang:

whoisj commented:
> @jamill :100: :bangbang:

Agreed. This is awesome - I am super excited to finally have rebase support. Oh the wonderful things we can bring to the world with this. :smiley:

paulgrattage commented:
Is it possible in git2sharp to rebase individual commits such as in this example http://blog.dennisrobinson.name/reorder-commits-with-git/ ? I am trying to do this in order to squash them into one commit which can then be cherrypicked.

dahlbyk commented:
If you just need to squash, you'll have an easier time with:

1. Soft reset (leave index intact) to merge base.
2. Commit.
3. Rebase/cherry-pick.

paulgrattage commented:
Thanks for that, this is actually what I have done at the moment. However there can be additional commits mixed in that I don't want to merge which was why I was looking at rebase. Alternatively if I could cherrypick multiple commits it would avoid the need to merge them at all but it seems Git2Sharp only supports picking 1 at a time.

ethomson commented:
> Alternatively if I could cherrypick multiple commits it would avoid the need to merge them at all but it seems Git2Sharp only supports picking 1 at a time.

I'm not sure that I understand what you mean.  You can only ever cherry-pick one at a time - git (for example) will simply apply them in serial.  (It doesn't create some strange multi-patch and try to apply them all at once).

So you should be able to just call cherry-pick in a loop...

dahlbyk commented:
Maybe referring to `cherry-pick -n`, which will apply multiple in serial but leaves all changes staged but uncommitted?

paulgrattage commented:
Thanks for that, I was considering using a loop but
git cherry-pick 51959c1 fc5f375 --no-commit or 
git cherry-pick 51959c1 --no-commit
git cherry-pick fc5f375 --no-commit and these work as expected. However if I use  
_repository.CherryPick(commit, author, new CherryPickOptions() { CommitOnSuccess = false }); on each of these commits the 2nd call throws  LibGit2Sharp.CheckoutConflictException  1 uncommitted change would be overwritten by merge
